base/Base.jl

```diff
@@ -275,24 +288,24 @@ include("set.jl")
 
 # Strings
 include("char.jl")
+function array_new_memory(mem::Memory{UInt8}, newlen::Int)
+    # add an optimization to array_new_memory for StringVector
+    if (@assume_effects :total @ccall jl_mem_owner(mem::Any,)::Any) isa String
+        # if data is in a String, keep it that way
+        # TODO: use jl_gc_expand_string(oldstr, newlen)?
+        str = _string_n(newlen)
+        return (@assume_effects :total !:consistent @ccall jl_string_to_genericmemory(str::Any,)::Memory{UInt8})
```

Comment:
is this really effect free?

Reply:
yes, allocation doesn't have side effects